src/bidiMapper/domains/context/logEntry.ts

```diff
@@ -0,0 +1,276 @@
+/**
+ * Copyright 2022 Google LLC.
+ * Copyright (c) Microsoft Corporation.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import { Protocol } from 'devtools-protocol';
+import { CdpClient } from '../../../cdp';
+import { CommonDataTypes, Log, Script } from '../../bidiProtocolTypes';
+import { IBidiServer } from '../../utils/bidiServer';
+import { IEventManager } from '../events/EventManager';
+
+export class LogEntry {
+  private specifiers = ['%s', '%d', '%i', '%f', '%o', '%O', '%c'];
+
+  private constructor(
+    private _contextId: string,
+    private _cdpClient: CdpClient,
+    private _bidiServer: IBidiServer,
+    private _eventManager: IEventManager,
+    private EVALUATOR_SCRIPT: string
+  ) {}
+
+  public static async create(
+    contextId: string,
+    cdpClient: CdpClient,
+    bidiServer: IBidiServer,
+    eventManager: IEventManager,
+    EVALUATOR_SCRIPT: string
+  ) {
+    const logEntry = new LogEntry(
+      contextId,
+      cdpClient,
+      bidiServer,
+      eventManager,
+      EVALUATOR_SCRIPT
+    );
+    await logEntry._initialize();
+    return logEntry;
+  }
+
+  private async _initialize() {
+    this._initializeEventListeners();
+  }
+
+  private _initializeEventListeners() {
+    this._initializeLogEntryAddedEventListener();
+  }
+
+  private _getLogLevel = (consoleAPIType: string): Log.LogLevel => {
+    if (consoleAPIType in ['error', 'assert']) {
+      return 'error';
+    }
+    if (consoleAPIType in ['debug', 'trace']) {
+      return 'debug';
+    }
+    if (consoleAPIType in ['warning', 'warn']) {
+      return 'warning';
+    }
+    return 'info';
+  };
+
+  private _isFormmatSpecifier(str: string): boolean {
+    return this.specifiers.some((spec) => str.includes(spec));
+  }
+
+  /**
+   * @param args input remote values to be format printed
+   * @returns parsed text of the remote values in specific format
+   */
+  private _formatter(args: CommonDataTypes.RemoteValue[]): string {
```

Comment:
I guess some unit tests here are necessary, as the logic is quite non-trivial.

Reply:
Added unit tests.